nl_server/config_reader.py

```diff
@@ -0,0 +1,215 @@
+      # a function.
+      if this_ver == 1:
+        # read indexes
+        for index_name, index_config in partial_catalog['indexes'].items():
+          store_type = index_config['store_type']
+          match store_type:
+            case StoreType.MEMORY:
+              indexes[index_name] = MemoryIndexConfig(**index_config)
+            case StoreType.LANCEDB:
+              indexes[index_name] = LanceDBIndexConfig(**index_config)
+            case StoreType.VERTEXAI:
+              indexes[index_name] = VertexAIIndexConfig(**index_config)
+            case _:
+              raise ValueError(f'Unknown store type: {store_type}')
+
+          # read models
+          for model_name, model_config in partial_catalog['models'].items():
+            model_type = model_config['type']
+            match model_type:
+              case ModelType.LOCAL:
+                models[model_name] = LocalModelConfig(**model_config)
+              case ModelType.VERTEXAI:
+                models[model_name] = VertexAIModelConfig(**model_config)
+              case _:
+                raise ValueError(f'Unknown model type: {model_type}')
+  logging.debug('NL Catalog:\n%s', json.dumps(asdict(catalog), indent=2))
+  return catalog
+
+
+def read_env() -> Env:
+  """Reads the envs
+  """
+  # TODO: Make this generic by checking a user provided config file path
+  if custom_dc_util.is_custom_dc():
+    c = Env.from_dict(yaml.safe_load(open(_ENV_USER_PATH)))
+  elif os.path.exists(_ENV_MOUNT_PATH):
+    c = Env.from_dict(yaml.safe_load(open(_ENV_MOUNT_PATH)))
+  else:
+    with open(_ENV_CODE_PATH) as f:
+      full_nl_config = yaml.safe_load(f.read())
+      c = Env.from_dict(full_nl_config['nl']['env'])
+  logging.debug('NL Env:\n%s', json.dumps(asdict(c), indent=2))
+  return c
+
+
+def get_server_config(catalog: Catalog, env: Env) -> ServerConfig:
+  """
+  Merges the catalog and env into a server config.
+  """
+  default_indexes = [x for x in env.default_indexes if x in catalog.indexes]
+
+  # Add reranking models
+  models = {}
+  if env.enable_reranking:
+    for model_name, model_config in catalog.models.items():
+      if model_config.usage == ModelUsage.RERANKING:
+        models[model_name] = model_config
+
+  # Only add enabled indexes
+  indexes = {}
+  for index_name in env.enabled_indexes:
+    if index_name not in catalog.indexes:
+      logging.warning('Index %s not found in catalog', index_name)
+      continue
+    index_config = catalog.indexes[index_name]
+    indexes[index_name] = index_config
+    model_name = index_config.model
+    models[model_name] = catalog.models[model_name]
+
+  # Add vertex AI model info
+  for model_name in models:
+    if model_name in env.vertex_ai_models:
+      m = models[model_name]
+      v = env.vertex_ai_models[model_name]
+      m_dict, v_dict = asdict(m), asdict(v)
+      merged = _merge_dicts(m_dict, v_dict)
+      models[model_name] = VertexAIModelConfig(**merged)
```

Comment:
i find it slight odd that the same dict starts with a dict and gets turned into ModelConfig.  should we have separate dicts with clearer types?

Reply:
the value is always a model config, so they were both converted to dict, merged, and then converted back to the class object. Added comment on preferring the env values.